src/v/serde/serde.h

```diff
@@ -0,0 +1,260 @@
+
+        auto size_placeholder = out.reserve(
+          sizeof(typename Type::envelope_size_t));
+        auto const size_before = out.size_bytes();
+
+        if constexpr (has_serde_write<Type>) {
+            t.serde_write(out);
+        } else {
+            envelope_for_each_field(t, [&out](auto& f) { write(out, f); });
+        }
+
+        auto const written_size = out.size_bytes() - size_before;
+        if (
+          written_size >
+          std::numeric_limits<typename Type::envelope_size_t>::max()) {
+            throw std::system_error{
+              make_error_code(error_codes::envelope_too_big)};
+        }
+        auto const size = ss::cpu_to_le(
+          static_cast<typename Type::envelope_size_t>(written_size));
+        size_placeholder.write(
+          reinterpret_cast<char const*>(&size),
+          sizeof(typename Type::envelope_size_t));
+    } else if constexpr (std::is_enum_v<Type>) {
+        write(out, static_cast<std::underlying_type_t<Type>>(t));
+    } else if constexpr (std::is_scalar_v<Type>) {
+        auto le_t = ss::cpu_to_le(t);
+        out.append(reinterpret_cast<char const*>(&le_t), sizeof(Type));
+    } else if constexpr (reflection::is_std_vector_v<Type>) {
+        assert(t.size() <= std::numeric_limits<int32_t>::max());
+        write(out, static_cast<int32_t>(t.size()));
+        for (auto const& el : t) {
+            write(out, el);
+        }
+    } else if constexpr (reflection::is_named_type_v<Type>) {
+        return write(iob, static_cast<typename Type::type>(t));
+    } else if constexpr (reflection::is_ss_bool_v<Type>) {
+        write(iob, static_cast<int8_t>(bool(t)));
+    } else if constexpr (std::is_same_v<Type, std::chrono::milliseconds>) {
+        write<int64_t>(iob, t.count());
+    } else if constexpr (std::is_same_v<Type, iobuf>) {
+        write<int32_t>(out, t.size_bytes());
+        out.append(std::move(t));
+    } else if constexpr (std::is_same_v<Type, ss::sstring>) {
+        write<int32_t>(out, t.size());
+        out.append(t.data(), t.size());
+    } else if constexpr (reflection::is_std_optional_v<Type>) {
+        if (t) {
+            write(out, true);
+            write(out, std::move(t.value()));
+        } else {
+            write(out, false);
+        }
+    }
+}
+
+template<typename T>
+std::decay_t<T> read(iobuf_parser&);
+
+template<typename T>
+header_t read_header(iobuf_parser& in) {
+    using Type = std::decay_t<T>;
+
+    auto const version = read<version_t>(in);
+    auto const compat_version = read<version_t>(in);
+    auto const size = read<typename Type::envelope_size_t>(in);
+
+    if (compat_version > Type::__version) {
+        serde_log.error(
+          "read compat_version={} > {}::version={}\n",
+          static_cast<int>(compat_version),
+          type_str<T>(),
+          static_cast<int>(Type::__version));
+        throw std::system_error{
+          make_error_code(error_codes::version_older_than_compat_version)};
```

Comment:
can you comment on logging the error here vs adding more context to the exception? I do agree though we do need some way to get contextual information out. also, I think we could probably do this trick here https://github.com/vectorizedio/redpanda/blob/dev/src/v/utils/utf8.h#L29. and with that trick it probably isn't so bad having the log statement.

Reply:
throw serde_exeception(fmt_with_ctx(ssx::format, "asdlkfjskldjf {} {}", params));